classes/plugins/DOIPubIdExportPlugin.inc.php

```diff
@@ -0,0 +1,411 @@
+	 * @param $object Issue|PublishedArticle
+	 * @return array (DAO, Method)
+	 */
+	function getDaoMethod($object) {
+		$configurations = array(
+				'Issue' => array('IssueDAO', 'updateObject'),
+				'Article' => array('ArticleDAO', 'updateObject'),
+		);
+		$foundConfig = false;
+		foreach($configurations as $objectType => $configuration) {
+			if (is_a($object, $objectType)) {
+				$foundConfig = true;
+				break;
+			}
+		}
+		assert($foundConfig);
+		return $configuration;
+	}
+
+	/**
+	 * Register the hook that adds an
+	 * additional field name to objects.
+	 * @param $daoName string
+	 */
+	function registerDaoHook($daoName) {
+		HookRegistry::register(strtolower_codesafe($daoName) . '::getAdditionalFieldNames', array(&$this, 'getAdditionalFieldNames'));
+	}
+
+	/**
+	 * Hook callback that returns the
+	 * "registeredDoi" setting's name prefixed with
+	 * the plug-in's id to avoid name collisions.
+	 * @see DAO::getAdditionalFieldNames()
+	 * @param $hookName string
+	 * @param $args array
+	 */
+	function getAdditionalFieldNames($hookName, $args) {
+		assert(count($args) == 2);
+		$dao =& $args[0];
+		$returner =& $args[1];
+		assert(is_array($returner));
+		$returner[] = $this->getPluginId() . '::' . DOI_EXPORT_REGISTERED_DOI;
+	}
+
+	/**
+	 * Check whether we are in test mode.
+	 * @param $request Request
+	 * @return boolean
+	 */
+	function isTestMode($request) {
+		return ($request->getUserVar('testMode') == '1');
+	}
+
+	/**
+	 * @see AcronPlugin::parseCronTab()
+	 */
+	function callbackParseCronTab($hookName, $args) {
+		$taskFilesPath =& $args[0];
+		$taskFilesPath[] = $this->getPluginPath() . DIRECTORY_SEPARATOR . 'scheduledTasks.xml';
+		return false;
+	}
+
+	/**
+	 * Retrieve all unregistered articles.
+	 * @param $context Journal
+	 * @return array
+	 */
+	function getUnregisteredArticles($context) {
+		// Retrieve all published articles that have not yet been registered.
+		$publishedArticleDao = DAORegistry::getDAO('PublishedArticleDAO'); /* @var $publishedArticleDao PublishedArticleDAO */
+		$articles = $publishedArticleDao->getByPubIdType(
+			$this->getPubIdType(),
+			$context?$context->getId():null,
+			null,
+			null,
+			null,
+			$this->getPluginId(). '::' . DOI_EXPORT_REGISTERED_DOI,
+			null,
+			null
+		);
+		return $articles->toArray();
+	}
+
+	/**
+	 * Retrieve all unregistered issues.
+	 * @param $context Journal
+	 * @return array
+	 */
+	function getUnregisteredIssues($context) {
+		// Retrieve all issues that have not yet been registered.
+		$issueDao = DAORegistry::getDAO('IssueDAO'); /* @var $issueDao IssueDAO */
+		$issues = $issueDao->getByPubIdType(
+			$this->getPubIdType(),
+			$context?$context->getId():null,
+			$this->getPluginId(). '::' . DOI_EXPORT_REGISTERED_DOI,
+			null,
+			null
+		);
+		// Cache issues.
+		$nullVar = null;
+		$cache = $this->getCache();
+		foreach ($issues as $issue) {
+			$cache->add($issue, $nullVar);
+			unset($issue);
+		}
+		return $issues->toArray();
```

Comment:
It looks like this code expects to iterate through the issue list twice, and I'm not sure that's possible (e.g. with a database result set). Better to convert to an array and then use the array in the foreach loop too.